src/dom_tree/tree.rs

```diff
@@ -0,0 +1,557 @@
+    /// It is possible to assign it to another node in the tree after this operation.
+    pub fn remove_from_parent(&self, id: &NodeId) {
+        let mut nodes = self.nodes.borrow_mut();
+        TreeNodeOps::remove_from_parent(nodes.deref_mut(), id);
+    }
+
+    #[deprecated(since = "0.10.0", note = "please use `insert_before_of` instead")]
+    /// Append a sibling node in the tree before the given node.
+    pub fn append_prev_sibling_of(&self, id: &NodeId, new_sibling_id: &NodeId) {
+        self.insert_before_of(id, new_sibling_id);
+    }
+
+    /// Append a sibling node in the tree before the given node.
+    pub fn insert_before_of(&self, id: &NodeId, new_sibling_id: &NodeId) {
+        let mut nodes = self.nodes.borrow_mut();
+        TreeNodeOps::insert_before_of(nodes.deref_mut(), id, new_sibling_id);
+    }
+
+    /// Append a sibling node in the tree after the given node.
+    pub fn insert_after_of(&self, id: &NodeId, new_sibling_id: &NodeId) {
+        let mut nodes = self.nodes.borrow_mut();
+        TreeNodeOps::insert_after_of(nodes.deref_mut(), id, new_sibling_id);
+    }
+
+    /// Changes the parent of children nodes of a node.
+    pub fn reparent_children_of(&self, id: &NodeId, new_parent_id: Option<NodeId>) {
+        let mut nodes = self.nodes.borrow_mut();
+        TreeNodeOps::reparent_children_of(nodes.deref_mut(), id, new_parent_id);
+    }
+
+    /// Detaches the children of a node.
+    pub fn remove_children_of(&self, id: &NodeId) {
+        self.reparent_children_of(id, None)
+    }
+}
+
+impl Tree {
+    /// Adds nodes from another tree to the current tree
+    pub(crate) fn merge(&self, other: Tree) {
+        // `parse_fragment` returns a document that looks like:
+        // <:root>                     id -> 0
+        //  <body>                     id -> 1
+        //      <html>                 id -> 2
+        //          things we need.
+        //      </html>
+        //  </body>
+        // <:root>
+        let mut nodes = self.nodes.borrow_mut();
+
+        let mut other_nodes = other.nodes.into_inner();
+
+        let offset = nodes.len();
+        let skip: usize = 3;
+        let id_offset = offset - skip;
+
+        for node in other_nodes.iter_mut().skip(skip) {
+            fix_node(node, id_offset);
+        }
+        nodes.extend(other_nodes.into_iter().skip(skip));
+    }
+
+    /// Get the new id, that is not in the Tree.
+    ///
+    /// This function doesn't add a new id.
+    /// it is just a convenient wrapper to get the new id.
+    pub(crate) fn get_new_id(&self) -> NodeId {
+        NodeId::new(self.nodes.borrow().len())
+    }
+
+    /// Adds nodes from another tree to the current tree and
+    /// then applies a function to the first  merged node
+    pub(crate) fn merge_with_fn<F>(&self, other: Tree, f: F)
+    where
+        F: FnOnce(NodeId),
+    {
+        let new_node_id = self.get_new_id();
+        self.merge(other);
+        f(new_node_id);
+    }
```

Comment:
It is correct. Because it skip N nodes from the other tree, and the first new node to add into current Tree, will have id that equal to the previous Tree size.